lib/rules/sort-destructure-keys.js

```diff
@@ -170,19 +144,76 @@ module.exports = {
     const sorter = createSorter(caseSensitive);
 
     return {
-      ObjectPattern(node) {
+      ObjectPattern(objectPatternNode) {
+        const scope = context.getScope();
+
+        function isReferencedByOtherProperties(id) {
+          for (const variable of scope.variables) {
+            if (variable.name !== id.name) {
+              continue;
+            }
+
+            for (const reference of variable.references) {
+              if (reference.identifier === id) {
+                continue;
+              }
+
+              let current = reference.identifier;
+              while (current) {
+                if (current === objectPatternNode) {
+                  return true;
+                }
+                current = current.parent;
+              }
+            }
+            return false;
+          }
+        }
+
+        /**
+         * Returns whether or not a node is safe to be sorted.
+         * @param {import('estree').ObjectPattern['properties'][number]} node
+         */
+        function shouldCheck(node) {
```

Comment:
Same request here: Let's make `shouldCheck` top-level, and it'll also need to take `scope` as a parameter.